lib/libvfio-user.c

```diff
@@ -927,126 +901,273 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages_in) + sizeof(*range_in) ||
-        dirty_pages_in->argsz > SERVER_MAX_DATA_XFER_SIZE ||
-        dirty_pages_in->argsz < sizeof(*dirty_pages_out)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size=%zu argsz=%u",
-                msg->in.iov.iov_len, dirty_pages_in->argsz);
-        return ERROR_INT(EINVAL);
+    if (msg->out.iov.iov_base == NULL) {
+        return ERROR_INT(ENOMEM);
     }
 
-    range_in = msg->in.iov.iov_base + sizeof(*dirty_pages_in);
+    memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+            sizeof(struct vfio_user_device_feature));
 
-    /*
-     * range_in is client-controlled, but we only need to protect against
-     * overflow here: we'll take MIN() against a validated value next, and
-     * dma_controller_dirty_page_get() will validate the actual ->bitmap.size
-     * value later, anyway.
-     */
-    argsz = satadd_u64(sizeof(*dirty_pages_out) + sizeof(*range_out),
-                       range_in->bitmap.size);
+    struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+
+    res->argsz = sizeof(struct vfio_user_device_feature)
+            + sizeof(struct vfio_user_device_feature_migration);
+
+    switch (req->flags & VFIO_DEVICE_FEATURE_MASK) {
+        case VFIO_DEVICE_FEATURE_MIGRATION: {
+            struct vfio_user_device_feature_migration *mig =
+                (void *)res->data;
+            // FIXME are these always supported? Can we consider to be
+            // "supported" if said support is just an empty callback?
+            //
+            // We don't need to return RUNNING or ERROR since they are
+            // always supported.
+            mig->flags = VFIO_MIGRATION_STOP_COPY
+                        | VFIO_MIGRATION_PRE_COPY;
+            return 0;
+        }
+
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE: {
+            struct vfio_user_device_feature_mig_state *state =
+                (void *)res->data;
+            state->device_state = migration_get_state(vfu_ctx);
+            return 0;
+        }
+        
+        default: return ERROR_INT(EINVAL);
+    }
+}
+
+static int
+handle_migration_device_feature_set(vfu_ctx_t *vfu_ctx, uint32_t feature,
+                                    struct vfio_user_device_feature *res)
+{
+    assert(feature == VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE);
+
+    struct vfio_user_device_feature_mig_state *state = (void *)res->data;
+
+    return migration_set_state(vfu_ctx, state->device_state);
+}
+
+static int
+handle_dma_device_feature_get(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
+                              struct vfio_user_device_feature *req)
+{
+    struct vfio_user_device_feature_dma_logging_report *rep =
+        (void *)req->data;
+
+    ssize_t bitmap_size = get_bitmap_size(rep->length, rep->page_size);
+    if (bitmap_size < 0) {
+        return bitmap_size;
+    }
+
+    msg->out.iov.iov_len = sizeof(struct vfio_user_device_feature)
```

Comment:
instead of repeating these two sizeof() multiple times, just do it once in a const var?